### PR TITLE
Setup Checkout (Update Payment Method)

### PR DESCRIPTION
**Pull request by zeyarpaing**

### CODE REVIEW

**Comment by zeyarpaing** [comment]
Docs Link: https://docs.getopenpay.com/docs/update-credit-card-payment-method

File: example/src/app/page.tsx [hunk 1]
```diff
@@ -14,15 +14,17 @@ import classNames from 'classnames';
 import { loadStripe } from '@stripe/stripe-js';
 
 type OnCheckoutSuccess = (invoiceUrls: string[], subscriptionIds: string[], customerId: string) => void;
-
+type OnSetupCheckoutSuccess = (paymentMethodId: string) => void;
```
Comment (syvlabs) [inline example/src/app/page.tsx:17]: Again let's rename this to "SetupPaymentMethodSuccess"

File: example/src/app/page.tsx [hunk 2]
```diff
@@ -14,15 +14,17 @@ import classNames from 'classnames';
...
 interface FormProps {
   token: string;
   separateFrames: boolean;
   onCheckoutSuccess: OnCheckoutSuccess;
+  onSetupCheckoutSuccess: OnSetupCheckoutSuccess;
```
Comment (syvlabs) [inline example/src/app/page.tsx:22]: Same

File: example/src/app/page.tsx [hunk 3]
```diff
@@ -14,15 +14,17 @@ import classNames from 'classnames';
...
+  baseUrl?: string;
 }
 
 const Form: FC<FormProps> = (props) => {
-  const { token, separateFrames, onCheckoutSuccess } = props;
+  const { token, separateFrames, onCheckoutSuccess, onSetupCheckoutSuccess } = props;
```
Comment (syvlabs) [inline example/src/app/page.tsx:27]: Same

File: example/src/app/page.tsx [hunk 4]
```diff
@@ -98,18 +92,24 @@ const Form: FC<FormProps> = (props) => {
       onValidationError={onValidationError}
       onCheckoutStarted={onCheckoutStarted}
       onCheckoutSuccess={onCheckoutSuccess}
+      onSetupCheckoutSuccess={(paymentMethodID) => {
```
Comment (syvlabs) [inline example/src/app/page.tsx:95]: Same for all of these

File: example/src/app/page.tsx [hunk 5]
```diff
@@ -98,18 +92,24 @@ const Form: FC<FormProps> = (props) => {
...
+        onSetupCheckoutSuccess(paymentMethodID);
+      }}
       onCheckoutError={onCheckoutError}
+      baseUrl={'http://localhost:3030'}
```
Comment (syvlabs) [inline example/src/app/page.tsx:99]: Is this `baseUrl` only for dev and stg? Not applicable to prod right?

**Review by syvlabs — CHANGES_REQUESTED** [review]

File: lib/components/form.tsx [hunk 6]
```diff
@@ -194,13 +195,20 @@ const ElementsForm: FC<ElementsFormProps> = (props) => {
...
       } else if (eventType === EventType.enum.CHECKOUT_SUCCESS) {
-        console.log('[form] Checkout complete:', eventPayload.invoiceUrls);
+        console.log('[form] Checkout complete:', eventPayload);
         setPreventClose(false);
         setTokenized(0);
         setCheckoutFired(false);
 
         if (onCheckoutSuccess)
           onCheckoutSuccess(eventPayload.invoiceUrls, eventPayload.subscriptionIds, eventPayload.customerId);
+      } else if (eventType === EventType.enum.SETUP_CHECKOUT_SUCCESS) {
```
Comment (syvlabs) [inline lib/components/form.tsx:205]: Can we rename to "setup payment method"? Hopefully its easy

File: lib/components/form.tsx [hunk 7]
```diff
@@ -304,12 +313,16 @@ const ElementsForm: FC<ElementsFormProps> = (props) => {
...
 
+  /**
+   * Check if all iframes have loaded and invoke `onLoad` callback
+   */
   useEffect(() => {
-    if (loaded || !formRef.current || !totalAmountAtoms) return;
+    if (loaded || !formRef.current) return;
 
-    if (iframes.length === Object.keys(eventTargets).length) {
+    if (iframes.length > 0 && iframes.length === Object.keys(eventTargets).length) {
```
Comment (syvlabs) [inline lib/components/form.tsx:322]: Why did we add the `> 0` check? There's a case when both `iframes.length` and eventTargets length are both 0?

**Review by syvlabs — CHANGES_REQUESTED** [review]

File: example/src/app/page.tsx (hunk 5 above)
Comment (zeyarpaing) [reply example/src/app/page.tsx:99]: sorry, it's just testing. forgot to remove.

File: example/src/app/page.tsx [hunk 8]
```diff
@@ -62,10 +55,10 @@ const Form: FC<FormProps> = (props) => {
     });
   };
 
-  const onLoad = (totalAmountAtoms: number, currency?: string): void => {
+  const onLoad = (totalAmountAtoms?: number, currency?: string): void => {
     setLoading(false);
     resetErrors();
-    setAmount(`${currency ? `${currency.toUpperCase()} ` : '$'}${totalAmountAtoms / 100}`);
+    if (totalAmountAtoms) setAmount(`${currency ? `${currency.toUpperCase()} ` : '$'}${totalAmountAtoms / 100}`);
```
Comment (syvlabs) [inline example/src/app/page.tsx:61]: Please avoid one-line `if`s

Should be

```
if (totalAmountAtoms) {
    setAmount(...)
}
```

File: example/src/app/page.tsx [hunk 9]
```diff
@@ -85,6 +78,7 @@ const Form: FC<FormProps> = (props) => {
 
   useEffect(() => {
     console.log(`Stripe JS can be loaded as a <script> in head (recommended), or loaded through ${loadStripe.name}`);
+    if (!token) return;
```
Comment (syvlabs) [inline example/src/app/page.tsx:81]: Same here

File: lib/components/form.tsx (hunk 7 above)
Comment (syvlabs) [reply lib/components/form.tsx:322]: Oh that's tricky, can you add a const to serve as a comment? something like

```
const areiFramesAllLoaded = ...

if (iFramesAllLoaded) {
...
}
```

File: example/src/app/page.tsx [hunk 10]
```diff
@@ -58,7 +58,9 @@ const Form: FC<FormProps> = (props) => {
   const onLoad = (totalAmountAtoms?: number, currency?: string): void => {
     setLoading(false);
     resetErrors();
-    if (totalAmountAtoms) setAmount(`${currency ? `${currency.toUpperCase()} ` : '$'}${totalAmountAtoms / 100}`);
+    if (totalAmountAtoms) {
+      setAmount(`${currency ? `${currency.toUpperCase()} ` : '$'}${totalAmountAtoms / 100}`);
```
Comment (syvlabs) [inline example/src/app/page.tsx:62]: Sorry one very last, please get the `atomToCurrency` from either CDE or Conn-to, then use it here instead of `/ 100`

**Review by syvlabs — APPROVED** [review]

**Pull request merged**